pilot/pkg/status/resourcelock.go

```diff
@@ -14,47 +14,168 @@
 
-import "sync"
+import (
+	"context"
+	"sync"
+	"time"
 
-type ResourceLock struct {
-	masterLock sync.RWMutex
-	listing    map[interface{}]*sync.Mutex
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	"istio.io/pkg/log"
+)
+
+// Task to be performed.
+type Task func(entry *cacheEntry) error
+
+// Worker queue implements an expandable goroutine pool which executes at most one concurrent routine per target
+// resource.  Multiple calls to Push() will not schedule multiple executions per target resource, but will ensure that
+// the single execution uses the latest value.
+type WorkerQueue interface {
+	// Push a task.
+	Push(target Resource, progress Progress)
+	// Run the loop until a signal on the context
+	Run(ctx context.Context)
+	// Delete a task
+	Delete(target lockResource)
 }
 
-func (r *ResourceLock) Lock(i interface{}) {
-	r.retrieveMutex(i).Lock()
+type queueImpl struct {
+	delay   time.Duration
+	tasks   []lockResource
+	cond    *sync.Cond
+	closing bool
+	work    Task
+
+	cacheLock   sync.Mutex
+	cache       map[lockResource]*cacheEntry
+	workerCount int
+	maxWorkers  int
 }
 
-func (r *ResourceLock) Unlock(i interface{}) {
-	r.retrieveMutex(i).Unlock()
+// NewQueue instantiates a queue with a processing function
+func NewQueue(errorDelay time.Duration, work Task, maxWorkers int) WorkerQueue {
+	return &queueImpl{
+		delay:       errorDelay,
+		tasks:       make([]lockResource, 0),
+		closing:     false,
+		cond:        sync.NewCond(&sync.Mutex{}),
+		work:        work,
+		workerCount: 0,
+		maxWorkers:  maxWorkers,
+		cache:       make(map[lockResource]*cacheEntry),
+	}
 }
 
-// returns value indicating if init was necessary
-func (r *ResourceLock) init() bool {
-	if r.listing == nil {
-		r.masterLock.Lock()
-		defer r.masterLock.Unlock()
-		// double check, per pattern
-		if r.listing == nil {
-			r.listing = make(map[interface{}]*sync.Mutex)
-		}
-		return true
+func (q *queueImpl) Push(target Resource, progress Progress) {
+	q.cacheLock.Lock()
+	defer q.cacheLock.Unlock()
+	key := convert(target)
+	_, inqueue := q.cache[key]
+	q.cache[key] = &cacheEntry{
+		cacheVal:      &target,
+		cacheProgress: &progress,
+	}
+	if !inqueue {
+		q.enqueue(key)
 	}
-	return false
+	q.maybeAddWorker()
 }
 
-func (r *ResourceLock) retrieveMutex(i interface{}) *sync.Mutex {
-	if !r.init() {
-		r.masterLock.RLock()
-		if result, ok := r.listing[i]; ok {
-			r.masterLock.RUnlock()
-			return result
+func (q *queueImpl) Delete(target lockResource) {
+	q.cacheLock.Lock()
+	defer q.cacheLock.Unlock()
+	delete(q.cache, target)
+}
+
+func (q *queueImpl) enqueue(item lockResource) {
+	q.cond.L.Lock()
+	defer q.cond.L.Unlock()
+	if !q.closing {
+		q.tasks = append(q.tasks, item)
+	}
+	q.cond.Signal()
+}
+
+func (q *queueImpl) maybeAddWorker() {
+	q.cond.L.Lock()
+	if q.workerCount >= q.maxWorkers {
+		q.cond.L.Unlock()
+		return
+	}
+	q.workerCount++
+	q.cond.L.Unlock()
+	go func() {
+		for {
+			q.cond.L.Lock()
+			for !q.closing && len(q.tasks) == 0 {
```

Comment:
One concern I have. say we have 1 worker. It is looping over this section. Now `enqueue` is called, triggering `maybeAddWorker`, which spawns another goroutine. Then one of the two goroutines will immediately exit (probably the new one, but not certain). So every enqueue will add an extremely short lived goroutine unless we are at capacity already (which I would expect is fairly rare).

I think this is purely a performance issue though, not correctness, and likely a small one at that

Reply:
I think this is still open

Reply:
I confirmed this is still happening